OsmAnd-java/src/main/java/net/osmand/router/RoutePlannerFrontEnd.java

```diff
@@ -851,103 +867,76 @@ public RouteCalcResult searchRoute(final RoutingContext ctx, LatLon start, LatLo
 			ctx.calculationProgress.requestPrivateAccessRouting = true;
 		}
 		if (hhRoutingConfig != null) {
-			if (ctx.nativeLib == null || hhRoutingType == HHRoutingType.JAVA) {
-				HHNetworkRouteRes r = runHHRoute(ctx, start, targets);
-				if ((r != null && r.isCorrect()) || useOnlyHHRouting) {
-					return r;
-				}
-			} else {
-				setStartEndToCtx(ctx, start, end, intermediates);
-				ctx.calculationProgress.nextIteration();
-				RouteCalcResult r = runNativeRouting(ctx, null, hhRoutingConfig);
-				ctx.calculationProgress.timeToCalculate = (System.nanoTime() - timeToCalculate);
-				//makeStartEndPointsPrecise no need, C++ does it
-				RouteResultPreparation.printResults(ctx, start, end, r.detailed);
-				if ((!r.detailed.isEmpty() && r.isCorrect()) || useOnlyHHRouting) {
-					return r;
+			ctx.calculationProgress.nextIteration();
+			HHNetworkRouteRes r = null;
+			if (ctx.nativeLib != null && hhRoutingType != HHRoutingType.JAVA) {
+				ctx.initStartEndPoints(start, end, intermediates);
+				RouteSegmentResult[] nr = runNativeRouting(ctx, hhRoutingConfig);
+				if (nr.length > 0) {
+					r = new HHNetworkRouteRes();
+					r.detailed.addAll(Arrays.asList(nr));
 				}
 			}
-		}
-		
-		double maxDistance = MapUtils.getDistance(start, end);
-		if (!intermediatesEmpty) {
-			LatLon b = start;
-			for (LatLon l : intermediates) {
-				maxDistance = Math.max(MapUtils.getDistance(b, l), maxDistance);
-				b = l;
+			r = runJavaHHRoute(ctx, start, targets);
```

Comment:
Perhaps else { here?
Otherwise after C++ routing we call Java routing

Reply:
Yes if routing result or  not correct we call further (but not necessary java) C++ as well?

Reply:
In this case, runJavaHHRoute() is always called after runNativeRouting(). Please fix.